File: src/js/core/diagram/code/traceval.js
define([
  "core/schema/code/factory",
  "core/semantics/code/interpreter",
  "core/expr/code/env",
  "core/expr/code/impl",
  "core/expr/code/eval",
  "core/schema/tools/union"
],
function(Factory, Interpreter, Env, Impl, Eval, Union) {
  var Traceval ;

  var TracevalCommand = MakeMixin([Interpreter.Dispatcher, Impl.EvalCommand], function() {
    this.eval_ETernOp = function(obj) {
      var self = this; 
      var e1, e2, e3, res, src;
      e1 = self.eval(obj.e1());
      e2 = self.eval(obj.e2());
      e3 = self.eval(obj.e3());
      res = e1
        ? e2
        : e3;
      src = self.$.D._get("factory")._get(obj.schema_class().name());
      src.set_e1(self.$.D._get("src")._get(obj.e1()));
      src.set_e2(self.$.D._get("src")._get(obj.e2()));
      src.set_e3(self.$.D._get("src")._get(obj.e3()));
      self.$.D._get("src")._set(obj, src);
      return res;
    };

    this.eval_EBinOp = function(obj) {
      var self = this; 
      var res, src;
      res = null;
      switch (obj.op()) {
        case "&":
          res = self.eval(obj.e1()) && self.eval(obj.e2());
          break;
        case "|":
          res = self.eval(obj.e1()) || self.eval(obj.e2());
          break;
        case "eql?":
          res = self.eval(obj.e1()) == self.eval(obj.e2());
          break;
        case "+":
          res = self.eval(obj.e1()) + self.eval(obj.e2());
          break;
        case "*":
          res = self.eval(obj.e1()) * self.eval(obj.e2());
          break;
        case "-":
          res = self.eval(obj.e1()) - self.eval(obj.e2());
          break;
        case "/":
          res = self.eval(obj.e1()) / self.eval(obj.e2());
          break;
        case "<":
          res = self.eval(obj.e1()) < self.eval(obj.e2());
          break;
        case ">":
          res = self.eval(obj.e1()) > self.eval(obj.e2());
          break;
        case "<=":
          res = self.eval(obj.e1()) <= self.eval(obj.e2());
          break;
        case ">=":
          res = self.eval(obj.e1()) >= self.eval(obj.e2());
          break;
        default:
          self.raise(S("Unknown operator (", obj.op(), ")"));
          break;
      }
      src = self.$.D._get("factory")._get(obj.schema_class().name());
      src.set_op(obj.op());
      src.set_e1(self.$.D._get("src")._get(obj.e1()));
      src.set_e2(self.$.D._get("src")._get(obj.e2()));
      self.$.D._get("src")._set(obj, src);
      return res;
    };

    this.eval_EUnOp = function(obj) {
      var self = this; 
      var res, src;
      res = obj.op() == "!"
        ? ! self.eval(obj.e())
        : self.raise(S("Unknown operator (", obj.op(), ")"));
      src = self.$.D._get("factory")._get(obj.schema_class().name());
      src.set_op(obj.op());
      src.set_e(self.$.D._get("src")._get(obj.e()));
      self.$.D._get("src")._set(obj, src);
      return res;
    };

    this.eval_EVar = function(obj) {
      var self = this; 
      var res, path;
      if (! self.$.D._get("env").has_key_P(obj.name())) {
        self.raise(S("ERROR: undefined variable ", obj.name(), " in ", self.$.D._get("env")));
      }
      res = self.$.D._get("env")._get(obj.name());
      if (! System.test_type(res, Factory.MObject)) {
        if (self.$.D._get("srctemp")._get(obj.name())) {
          self.$.D._get("src")._set(obj, self.$.D._get("srctemp")._get(obj.name()));
        } else {
          self.$.D._get("src")._set(obj, Eval.make_const(self.$.D._get("factory"), res));
        }
      } else {
        path = Union.Copy(self.$.D._get("factory"), res._path().path());
        self.$.D._get("src")._set(obj, path);
      }
      return res;
    };

    this.eval_EConst = function(obj) {
      var self = this; 
      var res;
      res = obj.val();
      self.$.D._get("src")._set(obj, Eval.make_const(self.$.D._get("factory"), res));
      return res;
    };

    this.eval_ENil = function(obj) {
      var self = this; 
      self.$.D._get("src")._set(obj, self.$.D._get("factory").ENil());
      return null;
    };

    this.eval_EField = function(obj) {
      var self = this; 
      var target, res, src;
      target = self.dynamic_bind(function() {
        return self.eval(obj.e());
      }, new EnsoHash ({ in_fc: false }));
      res = self.$.D._get("in_fc")
        ? target.method(obj.fname().to_sym())
        : target.send(obj.fname());
      src = self.$.D._get("factory")._get(obj.schema_class().name());
      src.set_fname(obj.fname());
      src.set_e(self.$.D._get("src")._get(obj.e()));
      self.$.D._get("src")._set(obj, src);
      return res;
    };

    this.eval_EFunCall = function(obj) {
      var self = this; 
      var m, params, clos, newsrctmp, param_src, res, b;
      m = self.dynamic_bind(function() {
        return self.eval(obj.fun());
      }, new EnsoHash ({ in_fc: true }));
      params = obj.params().map(function(p) {
        return self.eval(p);
      });
      if (obj.fun().EVar_P() && self.$.D._get("srctemp")._get(obj.fun().name()) != null) {
        clos = self.$.D._get("srctemp")._get(obj.fun().name());
        newsrctmp = self.$.D._get("srctemp").clone();
        clos.formals().each_with_index(function(f, i) {
          param_src = self.$.D._get("src")._get(obj.params()._get(i));
          return newsrctmp._set(f, param_src);
        });
        res = null;
        if (obj.lambda() == null) {
          self.dynamic_bind(function() {
            return res = m.apply(m, [].concat(params));
          }, new EnsoHash ({ srctemp: newsrctmp }));
        } else {
          b = self.eval(obj.lambda());
          self.dynamic_bind(function() {
            return res = m.apply(m, [b].concat(params));
          }, new EnsoHash ({ srctemp: newsrctmp }));
        }
        self.$.D._get("src")._set(obj, self.$.D._get("src")._get(clos.body()));
        return res;
      } else {
        if (obj.lambda() == null) {
          res = m.apply(m, [].concat(params));
        } else {
          b = self.eval(obj.lambda());
          res = m.apply(m, [b].concat(params));
        }
        self.$.D._get("src")._set(obj, Eval.make_const(self.$.D._get("factory"), res));
        return res;
      }
    };

    this.eval_EBlock = function(obj) {
      var self = this; 
      var res, defenv, env1, last;
      res = null;
      defenv = Env.HashEnv.new(new EnsoHash ({ }), self.$.D._get("env"));
      self.dynamic_bind(function() {
        return obj.fundefs().each(function(c) {
          return self.eval(c);
        });
      }, new EnsoHash ({ in_fc: false, env: defenv }));
      env1 = Env.HashEnv.new(new EnsoHash ({ }), defenv);
      self.dynamic_bind(function() {
        return obj.body().each(function(c) {
          return res = self.eval(c);
        });
      }, new EnsoHash ({ in_fc: false, env: env1 }));
      last = obj.body()._get(obj.body().size() - 1);
      self.$.D._get("src")._set(obj, self.$.D._get("src")._get(last));
      return res;
    };

    this.eval_EWhile = function(obj) {
      var self = this; 
      var res;
      res = ((function() {
        while (self.eval(obj.cond())) {
          self.eval(obj.body());
        }
      })());
      self.$.D._get("src")._set(obj, self.$.D._get("src")._get(obj.body()));
      return res;
    };

    this.eval_EFor = function(obj) {
      var self = this; 
      var nenv, res;
      nenv = Env.HashEnv.new(new EnsoHash ({ }), self.$.D._get("env"));
      res = self.eval(obj.list()).each(function(val) {
        nenv._set(obj.var(), val);
        return self.dynamic_bind(function() {
          return self.eval(obj.body());
        }, new EnsoHash ({ env: nenv }));
      });
      self.$.D._get("src")._set(obj, self.$.D._get("src")._get(obj.body()));
      return res;
    };

    this.eval_EIf = function(obj) {
      var self = this; 
      var res;
      if (self.eval(obj.cond())) {
        res = self.eval(obj.body());
        self.$.D._get("src")._set(obj, self.$.D._get("src")._get(obj.body()));
        return res;
      } else if (! (obj.body2() == null)) {
        res = self.eval(obj.body2());
        self.$.D._get("src")._set(obj, self.$.D._get("src")._get(obj.body2()));
        return res;
      }
    };

    this.eval_EFunDef = function(obj) {
      var self = this; 
      var forms;
      forms = [];
      obj.formals().each(function(f) {
        return forms.push(f.name());
      });
      self.$.D._get("env")._set(obj.name(), Impl.Closure.make_closure(obj.body(), forms, self.$.D._get("env"), self));
      self.$.D._get("srctemp")._set(obj.name(), Impl.Closure.new(obj.body(), forms, self.$.D._get("env"), self));
      return null;
    };

    this.eval_EAssign = function(obj) {
      var self = this; 
      self.lvalue(obj.var()).set_value(self.eval(obj.val()));
      if (obj.var().EVar_P()) {
        return self.$.D._get("srctemp")._set(obj.var().name(), self.$.D._get("src")._get(obj.val()));
      }
    };
  });

  var TracevalCommandC = MakeClass("TracevalCommandC", null, [TracevalCommand],
    function() {
    },
    function(super$) {
      this.initialize = function() {
        var self = this; 
      };
    });

  Traceval = {
    eval: function(obj, args) {
      var self = this; 
      if (args === undefined) args = new EnsoHash ({ });
      var interp;
      interp = TracevalCommandC.new();
      return interp.dynamic_bind(function() {
        return interp.eval(obj);
      }, args);
    },

    TracevalCommand: TracevalCommand,
    TracevalCommandC: TracevalCommandC,

  };
  return Traceval;
})

File: src/core/diagram/code/traceval.rb
require 'core/schema/code/factory'
require 'core/semantics/code/interpreter'
require 'core/expr/code/env'
require 'core/expr/code/impl'
require 'core/expr/code/eval'
require 'core/schema/tools/union'

module Traceval
  module TracevalCommand
    include Interpreter::Dispatcher
    include Impl::EvalCommand

    def eval_EBinOp(obj)
      res = nil
      case obj.op
      when "&"
        res = eval(obj.e1) && eval(obj.e2)
      when "|"
        res = eval(obj.e1) || eval(obj.e2)
      when "eql?"
        res = eval(obj.e1) == eval(obj.e2)
      when "+"
        res = eval(obj.e1) + eval(obj.e2)
      when "*"
        res = eval(obj.e1) * eval(obj.e2)
      when "-"
        res = eval(obj.e1) - eval(obj.e2)
      when "/"
        res = eval(obj.e1) / eval(obj.e2)
      when "<"
        res = eval(obj.e1) < eval(obj.e2)
      when ">"
        res = eval(obj.e1) > eval(obj.e2)
      when "<=" 
        res = eval(obj.e1) <= eval(obj.e2)
      when ">=" 
        res = eval(obj.e1) >= eval(obj.e2)
      else
        raise "Unknown operator (#{obj.op})"
      end
      src = @D[:factory][obj.schema_class.name]
      src.op = obj.op
      src.e1 = @D[:src][obj.e1]
      src.e2 = @D[:src][obj.e2]
      @D[:src][obj] = src
      res
    end

    def eval_EUnOp(obj)
      res = if obj.op == "!"
        !eval(obj.e)
      else
        raise "Unknown operator (#{obj.op})"
      end
      src = @D[:factory][obj.schema_class.name]
      src.op = obj.op
      src.e = @D[:src][obj.e]
      @D[:src][obj] = src
      res
    end
  
    def eval_EVar(obj)
      raise "ERROR: undefined variable #{obj.name} in #{@D[:env]}" unless @D[:env].has_key?(obj.name)
      res = @D[:env][obj.name]
      if !res.is_a? Factory::MObject #if this result is a constant then just list it as a constant
        if @D[:srctemp][obj.name]
          @D[:src][obj] = @D[:srctemp][obj.name]
        else
          @D[:src][obj] = Eval::make_const(@D[:factory], res)
        end
      else #if it came from somewhere then remember that place as a path, ie EField
        path = Union::Copy(@D[:factory], res._path.path)
        @D[:src][obj] = path
      end
      res
    end

    def eval_EConst(obj)
      res = obj.val
      @D[:src][obj] = Eval::make_const(@D[:factory], res)
      res
    end

    def eval_ENil(obj)
      @D[:src][obj] = @D[:factory].ENil
      nil
    end

    def eval_EField(obj)
      target = dynamic_bind in_fc: false do
        eval(obj.e)
      end
      res = if @D[:in_fc]
        target.method(obj.fname.to_sym)
      else
        target.send(obj.fname)
      end
      src = @D[:factory][obj.schema_class.name]
      src.fname = obj.fname
      src.e = @D[:src][obj.e]
      @D[:src][obj] = src
      res
    end

    def eval_EFunCall(obj)
      m = dynamic_bind in_fc: true do
        eval(obj.fun)
      end
      params = obj.params.map{|p|eval(p)}
      if obj.fun.EVar? and @D[:srctemp][obj.fun.name]!=nil
        clos = @D[:srctemp][obj.fun.name]
        newsrctmp = @D[:srctemp].clone
        clos.formals.each_with_index do |f,i|
          param_src = @D[:src][obj.params[i]]
          newsrctmp[f] = param_src
        end
        res = nil
        if obj.lambda.nil?
          dynamic_bind srctemp: newsrctmp do
            res = m.call(*params)
          end
        else
          b = eval(obj.lambda)
          dynamic_bind srctemp: newsrctmp do
            res = m.call(*params, &b)
          end
        end
        @D[:src][obj] = @D[:src][clos.body]
        res
      else
        if obj.lambda.nil?
          res = m.call(*params)
        else
          b = eval(obj.lambda)
          res = m.call(*params, &b)
        end
        @D[:src][obj] = Eval::make_const(@D[:factory], res)
        res
      end
    end
    
    def eval_EBlock(obj)
      res = nil
      #fundefs are able to see each other but not any other variable created in the block
      defenv = Env::HashEnv.new({}, @D[:env])
      dynamic_bind in_fc: false, env: defenv do
        obj.fundefs.each do |c|
          eval(c)
        end
      end
      #rest of body can see fundefs
      env1 = Env::HashEnv.new({}, defenv)
      dynamic_bind in_fc: false, env: env1 do
        obj.body.each do |c|
          res = eval(c)
        end
      end
      last = obj.body[obj.body.size-1]
      @D[:src][obj] = @D[:src][last]
      res
    end

    def eval_EWhile(obj)
      res = while eval(obj.cond)
        eval(obj.body)
      end
      @D[:src][obj] = @D[:src][obj.body]
      res
    end

    def eval_EFor(obj)
      nenv = Env::HashEnv.new({obj.var=>nil}, @D[:env])
      res = eval(obj.list).each do |val|
        nenv[obj.var] = val
        dynamic_bind env: nenv do
          eval(obj.body)
        end
      end
      @D[:src][obj] = @D[:src][obj.body]
      res
    end

    def eval_EIf(obj)
      if eval(obj.cond)
        res = eval(obj.body)
        @D[:src][obj] = @D[:src][obj.body]
        res
      elsif !obj.body2.nil?
        res = eval(obj.body2)
        @D[:src][obj] = @D[:src][obj.body2]
        res
      end
    end

    def eval_EFunDef(obj)
      forms = []
      obj.formals.each {|f| forms << f.name}
      @D[:env][obj.name] = Impl::Closure.make_closure(obj.body, forms, @D[:env], self)
      @D[:srctemp][obj.name] = Impl::Closure.new(obj.body, forms, @D[:env], self)
      nil
    end

    def eval_EAssign(obj)
      lvalue(obj.var).value = eval(obj.val)
      if (obj.var.EVar?)
        @D[:srctemp][obj.var.name] = @D[:src][obj.val]
      end
    end
  end

  class TracevalCommandC
    include TracevalCommand
    def initialize; end
  end

  def self.eval(obj, args={})
    interp = TracevalCommandC.new
    interp.dynamic_bind args do
      interp.eval(obj)
    end
  end

end
